src/core/inductive.ml

```diff
@@ -55,17 +61,67 @@ let gen_ind_typ_codom : popt -> sym -> (tbox list -> tbox) -> string -> tbox =
     | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
   in aux [] !(ind_sym.sym_type)
 
+(** [prf_of p c ts t] returns a proof of the form
+    c.symb_prf ( (((p ts1) ...) tsn) t) where ts = [ts1;...;tsn]. *)
+let prf_of : tvar -> config -> tbox list -> tbox -> tbox = fun p c ts t ->
+  _Appl_symb c.symb_prf [_Appl (_Appl_list (_Vari p) ts) t]
+
+(** [preprocessing pos c ind_typ_list p_str x_str] computes elements of the
+    induction principles that we want to associate to a list of mutually
+    defined inductive types [ind_typ_list]. It returns a pair of lists:
+      - The first list is an association list mapping every inductive type to
+        a pair made of a predicate variable and its type.The name of a
+        predicate variable has the type tvar and its type has the type tbox.
+      - The second list is composed of the conclusions of the induction
+      principles.
+    The string [p_str] is used as prefix for predicate variables, and the
+    string [x_str] as prefix for type parameters. *)
+let preprocessing :
+      popt -> config -> sym list -> string -> string
+      -> ((sym * (tvar * tbox)) list * tbox list) =
```

Comment:
Is it the case that the two output lists have the same length? In this case, why not returning a single list of type `(sym * (tvar * tbox * tbox)) list`?

Reply:
Or rather `(sym * (tvar * (tbox * tbox))) list` (see comment about fold_cons_typ).

Reply:
You could give a name to this type like `sym_pred_map`.

Reply:
This clarifies the goal of preprocessing: building this map from the list of type symbols. Perhaps, you could rename it as create_sym_pred_map.

Reply:
> Is it the case that the two output lists have the same length? In this case, why not returning a single list of type `(sym * (tvar * tbox * tbox)) list`?

Yes these lists have the same size but are never used together. I find it ambiguous to link these lists, and therefore not very readable.

> Or rather `(sym * (tvar * (tbox * tbox))) list` (see comment about fold_cons_typ).

I don't understand.






> You could give a name to this type like `sym_pred_map`.

I don't understand, the type `(sym * (tvar * (tbox * tbox))) list` used once.



> This clarifies the goal of preprocessing: building this map from the list of type symbols. Perhaps, you could rename it as create_sym_pred_map.

I don't understand, the goal is clearly written.
I don't find your name very clear, unlike 'preprocessing'.

Reply:
It's not because a type is used only once that it does not deserve a definition. This makes things more readable. It's like a let in a piece of code.
Preprocessing is too generic; it doesn't mean much.
It's not because these two lists are not used together that they cannot be merged.

Reply:
> It's not because a type is used only once that it does not deserve a definition. This makes things more readable. It's like a let in a piece of code.

I prefer `sym_predicat_map`.
> Preprocessing is too generic; it doesn't mean much.

Yes, but this function doesn't just return a map.
> It's not because these two lists are not used together that they cannot be merged.

Why would you like to merge the predicate map and the list of conclusions?
This change doesn't avoid any invariant.

Reply:
There is a conclusion for each inductive type. So it makes sense to put this into a single map mapping to every inductive type all the data related to each type.

Reply:
Dans c'est cas là pourquoi ne pas aussi associer une règle de réécriture à sa clause et à son constructeur ?
On aurait un unique fold, qui fait tout, et qui n'a pas besoin d'être instancié, car de toute façon on ne l'instanciera pas plus de 2 fois.

Je ne comprends pas pourquoi vouloir tout regrouper comme cela. C'était justifié avec ces histoires d'invariants à maintenir, mais à présent je trouve que cela complique tout.

Je ne me sens pas capable de maintenir un truc pareil par la suite.